File: app/src/main/java/es/etologic/mahjongscoring2/app/combinations/CombinationsRvAdapter.java
package es.etologic.mahjongscoring2.app.combinations;

import android.animation.ValueAnimator;
import android.content.Context;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import es.etologic.mahjongscoring2.R;
import es.etologic.mahjongscoring2.app.model.ShowState;
import es.etologic.mahjongscoring2.domain.entities.Combination;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static es.etologic.mahjongscoring2.app.model.ShowState.HIDE;
import static es.etologic.mahjongscoring2.app.model.ShowState.SHOW;
import static es.etologic.mahjongscoring2.domain.entities.Combination.CombinationDescriptionType.IMAGE;

class CombinationsRvAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    //region Constants

    private static final int ANIMATION_DURATION = 500;

    //endregion

    //region Fields

    private List<Combination> combinations;
    private ShowState imageOrDescriptionShowState;
    private int cardViewFullHeight;
    private int cardViewMinHeight;

    //endregion

    //region Constructor

    CombinationsRvAdapter() {
        combinations = new ArrayList<>();
        imageOrDescriptionShowState = HIDE;
    }

    //endregion

    //region Public

    void setCombinations(List<Combination> combinations) {
        saveCombinationsCopy(combinations);
        notifyDataSetChanged();
    }

    ShowState toggleImageOrDescription() {
        imageOrDescriptionShowState = imageOrDescriptionShowState == SHOW ? HIDE : SHOW;
        notifyDataSetChanged();
        return imageOrDescriptionShowState;
    }

    //endregion

    //region ViewHolder

    class CombinationItemViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.llCombinationItemContainer) LinearLayout llContainer;
        @BindView(R.id.cvCombinationItem) CardView cardView;
        @BindView(R.id.tvCombinationItemPoints) TextView tvPoints;
        @BindView(R.id.tvCombinationItemName) TextView tvName;
        @BindView(R.id.tvCombinationItemPosition) TextView tvPosition;
        @BindView(R.id.flCombinationItemImageOrDescriptionContainer) FrameLayout flImageOrDescriptionContainer;
        @BindView(R.id.ivCombinationItemImage) ImageView ivImage;
        @BindView(R.id.tvCombinationItemDescription) TextView tvDescription;

        CombinationItemViewHolder(View view) {
            super(view);
            ButterKnife.bind(this, view);
        }

        @OnClick(R.id.cvCombinationItem) void onCombinationItemClick() {
            if (cardView.getHeight() == cardViewMinHeight) {
                showImageOrDescriptionAnimated(this);
            } else if(cardView.getHeight() == cardViewFullHeight){
                hideImageOrDescriptionAnimated(this);
            }
        }
    }

    //endregion

    //region Lifecycle

    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
        Context context = recyclerView.getContext();
        cardViewMinHeight = (int) context.getResources().getDimension(
                R.dimen.combination_item_cardview_min_height);
        cardViewFullHeight = (int) context.getResources().getDimension(
                R.dimen.combination_item_cardview_full_height);
    }

    @Override
    public int getItemCount() {
        return combinations.size();
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.combination_item,
                parent, false);
        return new CombinationItemViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        Combination combination = combinations.get(position);
        CombinationItemViewHolder myHolder = (CombinationItemViewHolder) holder;
        //region Points
        myHolder.tvPoints.setText(String.valueOf(combination.getCombinationPoints()));
        //endregion
        //region Name
        myHolder.tvName.setText(combination.getCombinationName());
        //endregion
        //region Position
        myHolder.tvPosition.setText(String.format(Locale.getDefault(),"#%d", position+1));
        //endregion
        //region Image or Description
        if(imageOrDescriptionShowState == SHOW) {
            if(myHolder.cardView.getHeight() == cardViewMinHeight) {
                showImageOrDescriptionNoAnimated(myHolder);
            }
        } else if(myHolder.cardView.getHeight() == cardViewFullHeight) {
            hideImageOrDescriptionNoAnimated(myHolder);
        }

        if(combination.getCombinationDescriptionType() == IMAGE) {
            myHolder.ivImage.setImageResource(combination.getCombinationImage());
            myHolder.tvDescription.setVisibility(GONE);
            myHolder.ivImage.setVisibility(VISIBLE);
        } else {
            myHolder.tvDescription.setText(combination.getCombinationDescription());
            myHolder.ivImage.setVisibility(GONE);
            myHolder.tvDescription.setVisibility(VISIBLE);
        }
        //endregion
    }

    //endregion

    //region Private

    private void saveCombinationsCopy(List<Combination> combinations) {
        List<Combination> newCombinationsCopy = new ArrayList<>(combinations.size());
        for(Combination combination : combinations) {
            newCombinationsCopy.add(combination.getCopy());
        }
        this.combinations = newCombinationsCopy;
    }

    private void showImageOrDescriptionNoAnimated(CombinationItemViewHolder myHolder) {
        myHolder.flImageOrDescriptionContainer.setVisibility(VISIBLE);
    }

    private void hideImageOrDescriptionNoAnimated(CombinationItemViewHolder myHolder) {
        myHolder.flImageOrDescriptionContainer.setVisibility(GONE);
    }

    private void showImageOrDescriptionAnimated(CombinationItemViewHolder holder) {
        ValueAnimator expandAnimation = ValueAnimator.ofInt(
                holder.cardView.getMeasuredHeightAndState(), cardViewFullHeight);
        expandAnimation.addUpdateListener(valueAnimator -> {
            int val = (Integer)valueAnimator.getAnimatedValue();
            ViewGroup.LayoutParams layoutParams = holder.cardView.getLayoutParams();
            layoutParams.height = val;
            holder.cardView.setLayoutParams(layoutParams);
        });
        expandAnimation.setDuration(ANIMATION_DURATION);
        AlphaAnimation fadeInAnimation = new AlphaAnimation(0, 1);
        fadeInAnimation.setDuration(ANIMATION_DURATION);
        fadeInAnimation.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {
                showImageOrDescriptionNoAnimated(holder);
            }

            @Override
            public void onAnimationEnd(Animation animation) {}

            @Override
            public void onAnimationRepeat(Animation animation) {}
        });
        expandAnimation.start();
        holder.flImageOrDescriptionContainer.startAnimation(fadeInAnimation);
    }

    private void hideImageOrDescriptionAnimated(CombinationItemViewHolder holder) {
        ValueAnimator collapseAnimation = ValueAnimator.ofInt(
                holder.cardView.getMeasuredHeightAndState(), cardViewMinHeight);
        collapseAnimation.addUpdateListener(valueAnimator -> {
            int val = (Integer)valueAnimator.getAnimatedValue();
            ViewGroup.LayoutParams layoutParams = holder.cardView.getLayoutParams();
            layoutParams.height = val;
            holder.cardView.setLayoutParams(layoutParams);
        });
        collapseAnimation.start();
        collapseAnimation.setDuration(ANIMATION_DURATION);
        AlphaAnimation fadeOutAnimation = new AlphaAnimation(1, 0);
        fadeOutAnimation.setDuration(ANIMATION_DURATION);
        fadeOutAnimation.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {}

            @Override
            public void onAnimationEnd(Animation animation) {
                hideImageOrDescriptionNoAnimated(holder);
            }

            @Override
            public void onAnimationRepeat(Animation animation) {}
        });
        holder.flImageOrDescriptionContainer.startAnimation(fadeOutAnimation);
        collapseAnimation.start();
    }

    //endregion
}

File: app/src/main/java/com/etologic/mahjongscoring2/app/main/combinations/CombinationsRvAdapter.kt
/*
    Copyright (C) 2020  Ernesto Vega de la Iglesia Soria
 
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
    
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    
    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
package com.etologic.mahjongscoring2.app.main.combinations

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import com.etologic.mahjongscoring2.R
import com.etologic.mahjongscoring2.app.extensions.setOnSecureClickListener
import com.etologic.mahjongscoring2.app.model.ShowState
import com.etologic.mahjongscoring2.app.model.ShowState.HIDE
import com.etologic.mahjongscoring2.app.model.ShowState.SHOW
import com.etologic.mahjongscoring2.business.model.entities.Combination
import com.etologic.mahjongscoring2.business.model.entities.Combination.CombinationDescriptionType.IMAGE
import com.etologic.mahjongscoring2.databinding.CombinationItemBinding
import java.util.*
import javax.inject.Inject

internal class CombinationsRvAdapter
@Inject constructor() : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    
    private var combinations: List<Combination> = ArrayList()
    private var imageOrDescriptionShowState = HIDE
    private var cardViewMinHeight: Int = 0
    
    @SuppressLint("NotifyDataSetChanged")
    internal fun setCombinations(combinations: List<Combination>) {
        saveCombinationsCopy(combinations)
        notifyDataSetChanged()
    }
    
    private fun saveCombinationsCopy(combinations: List<Combination>) {
        val newCombinationsCopy = ArrayList<Combination>(combinations.size)
        for (combination in combinations) {
            newCombinationsCopy.add(combination.getCopy())
        }
        this.combinations = newCombinationsCopy
    }
    
    fun toggleImageOrDescription(): ShowState {
        imageOrDescriptionShowState = if (imageOrDescriptionShowState === SHOW) HIDE else SHOW
        notifyItemRangeChanged(0, combinations.size - 1)
        return imageOrDescriptionShowState
    }
    
    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        cardViewMinHeight = recyclerView.context.resources.getDimension(
            R.dimen.combination_item_cardview_min_height
        ).toInt()
    }
    
    override fun getItemCount(): Int {
        return combinations.size
    }
    
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val itemBinding = CombinationItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return CombinationItemViewHolder(itemBinding)
    }
    
    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val combination = combinations[position]
        val myHolder = holder as CombinationItemViewHolder
        myHolder.tvPoints.text = combination.combinationPoints.toString()
        myHolder.tvName.text = combination.combinationName
        myHolder.tvPosition.text = String.format(Locale.getDefault(), "#%d", position + 1)
        
        if (combination.combinationDescriptionType === IMAGE) {
            myHolder.ivImage.setImageResource(combination.combinationImage)
            myHolder.tvDescription.visibility = GONE
            myHolder.ivImage.visibility = VISIBLE
        } else {
            myHolder.tvDescription.text = combination.combinationDescription
            myHolder.ivImage.visibility = GONE
            myHolder.tvDescription.visibility = VISIBLE
        }
        
        myHolder.flImageOrDescriptionContainer.visibility = if (imageOrDescriptionShowState === SHOW) VISIBLE else GONE
        
        myHolder.llContainer.setOnSecureClickListener {
            myHolder.flImageOrDescriptionContainer.visibility = if (myHolder.cardView.height == cardViewMinHeight) VISIBLE else GONE
        }
    }
    
    internal inner class CombinationItemViewHolder(binding: CombinationItemBinding) : RecyclerView.ViewHolder(binding.root) {
        
        val llContainer: LinearLayout = binding.llCombinationItemContainer
        val cardView: CardView = binding.cvCombinationItem
        val tvPoints: TextView = binding.tvCombinationItemPoints
        val tvName: TextView = binding.tvCombinationItemName
        val tvPosition: TextView = binding.tvCombinationItemPosition
        val flImageOrDescriptionContainer: FrameLayout = binding.flCombinationItemImageOrDescriptionContainer
        val ivImage: ImageView = binding.ivCombinationItemImage
        val tvDescription: TextView = binding.tvCombinationItemDescription
    }
}